extensions/control-plane/provision/provision-aws-s3-copy/src/main/java/org/eclipse/edc/connector/provision/aws/s3/copy/S3CopyDeprovisionPipeline.java

```diff
@@ -0,0 +1,215 @@
+/*
+ *  Copyright (c) 2025 Cofinity-X
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Cofinity-X - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.connector.provision.aws.s3.copy;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import jakarta.json.Json;
+import jakarta.json.JsonObject;
+import org.eclipse.edc.aws.s3.AwsClientProvider;
+import org.eclipse.edc.aws.s3.S3ClientRequest;
+import org.eclipse.edc.connector.controlplane.transfer.spi.types.DeprovisionedResource;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.security.Vault;
+import org.eclipse.edc.spi.types.TypeManager;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.iam.IamAsyncClient;
+import software.amazon.awssdk.services.iam.model.DeleteRolePolicyRequest;
+import software.amazon.awssdk.services.iam.model.DeleteRolePolicyResponse;
+import software.amazon.awssdk.services.iam.model.DeleteRoleRequest;
+import software.amazon.awssdk.services.iam.model.DeleteRoleResponse;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.DeleteBucketPolicyRequest;
+import software.amazon.awssdk.services.s3.model.GetBucketPolicyRequest;
+import software.amazon.awssdk.services.s3.model.GetBucketPolicyResponse;
+import software.amazon.awssdk.services.s3.model.PutBucketPolicyRequest;
+import software.amazon.awssdk.services.s3.model.S3Response;
+
+import java.util.HashMap;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import static org.eclipse.edc.aws.s3.copy.lib.S3CopyUtils.getSecretTokenFromVault;
+import static org.eclipse.edc.connector.provision.aws.s3.copy.util.S3CopyProvisionConstants.S3_BUCKET_POLICY_STATEMENT;
+import static org.eclipse.edc.connector.provision.aws.s3.copy.util.S3CopyProvisionConstants.S3_BUCKET_POLICY_STATEMENT_SID;
+
+/**
+ * Deprovisions the AWS resources and policies added through the provisiong process before a
+ * cross-account copy of S3 objects. This includes resetting the bucket policy in the destination
+ * account and deleting the dedicated role in the source account. All operations are executed
+ * consecutively.
+ */
+public class S3CopyDeprovisionPipeline {
```

Comment:
I see that this design comes by the current `S3DeprovisionPipeline`, but the be honest I never bore with this way, if the goal is to not to have too code in the provisioner I would tackle differently, e.g. one thing that could be done is to have wrappers around the aws clients that could provide more high level features like `WrapperIamClient.deleteRole`, this will make this code also available to the other provisioning/transfer service modules, isolate aws sdk apis, provide better testability and improve the readability of the provisioner code, that it could resolve in some meaningful lines instead of just relying on a "pipeline" component that looks like a carpet where all the dirt is swept.

No request to any change here, but food for thoughts for potential upcoming refactoring

Reply:
Totally agree. Initially I liked this design with the pipelines, but they ended up getting quite unreadable and hard to test. Will create an issue for that once the PR is merged